[wip] Add `InstanceName` and `InstanceDirName` functions

anjannath opened the pull request:
The instance directory and instance name for libmachine is same but
for better clearity two funcs InstanceName() and  InstanceDirName()
are added

When we are dealing with the VM name we should use InstanceName and
for dealing with the paths for the VM we should use InstanceDirName

_This is initial work to have the instance name based on the preset, so that multiple instances can exist simultaneously. Later we'll modify the `InstanceName` funtion to return the name based on the preset_

anjannath committed:
Differentiate between instance name and instance directory

The instance directory and instance name for libmachine is same but
for better clearity two funcs InstanceName() and  InstanceDirName()
are added

When we are dealing with the VM name we are using InstanceName and
for dealing with the paths for the VM we are using InstanceDirName

anjannath committed:
Use InstanceName() and InstanceDirName() in place of DefaultName

Replace use of constants.DefaultName in code dealing with VM name with
InstanceName() and code dealing with VM file paths with InstanceDirName()

anjannath committed:
The cluster name is independent of instance dir or instance name

renamed DefaultName to DefaultClusterName which is always "crc" and
that is set in the bundles

anjannath committed:
libvirt: don't fix the storage pool name to 'crc'

if this is not set then the driver will use the driver.Name as  the pool  name
which will be useful as it'll use different storage pool for differently named
virtual machines

praveenkumar commented:
I am expecting as of now this PR is base work for having different directory for different preset, can we add more commits to it (even for default preset) to provide sample logic how these changes are consumed because as of now we are just adding 2 helper functions and refactoring code around it but doesn't provide an idea how it is consumable from preset context (even bit details on the issue or here might help to understand the usecase of it).

anjannath commented:
> I am expecting as of now this PR is base work for having different directory for different preset, can we add more commits to it (even for default preset) to provide sample logic how these changes are consumed because as of now we are just adding 2 helper functions and refactoring code around it but doesn't provide an idea how it is consumable from preset context (even bit details on the issue or here might help to understand the usecase of it).

my plan is to next modify `InstanceName` to return the `crc-$preset` as the name, to have the instances differently named, i'll add more commits to the PR

anjannath committed:
Make InstanceDirName and InstanceName dependent on the Preset

this updates all the functions that use the InstanceName or the
InstanceDirName functions to determine the name and path of the
VM disk image

jsliacan commented on test/integration/utilities_test.go:
```diff
@@ -97,8 +98,8 @@ func RunCRCExpectFail(args ...string) (string, error) {
-func SendCommandToVM(cmd string) (string, error) {
-	client := machine.NewClient(constants.InstanceName(), false,
+func SendCommandToVM(cmd string, preset crcPreset.Preset) (string, error) {
+	client := machine.NewClient(constants.InstanceName(preset), false,
```
I see, so this `constants.InstanceName` is now preset dependent, and the changes we need in tests are all either directly to the `constants.InstanceName` or to the `SendCommandToVM` function which utilizes this. LGTM. Thanks for the changes!

anjannath committed:
Fix path to docker socket and ssh key for podman preset test

anjannath committed:
api: update machine client with updated instance name

the machine client that the http api handlers have access to
is passed from the daemon command package and is reused  for
the entire daemon run, this works as long as there's  single
VM. When there's more then one VM and with different VM name
we need to re-create the machine client with the correct  VM
name which depends on the current preset

this adds a field to the Handler struct 'MachineClientUpdater'
which is a function that returns a new machine.Client for the
correct instance name returned by constants.InstanceName

anjannath committed:
api: add mock http handler with fake MachineClientUpdater for tests

gbraad commented:
@praveenkumar what more do you want to know?

cfergeau committed:
Remove unused UseOCWithConfig

cfergeau committed:
Introduce constants.ResolveInstancePath()

cfergeau commented on pkg/crc/constants/constants.go:
```diff
@@ -109,8 +109,8 @@ var (
 	DaemonSocketPath   = filepath.Join(CrcBaseDir, "crc.sock")
-	KubeconfigFilePath = filepath.Join(MachineInstanceDir, DefaultName, "kubeconfig")
-	PasswdFilePath     = filepath.Join(MachineInstanceDir, DefaultName, "passwd")
+	KubeconfigFilePath = filepath.Join(MachineInstanceDir, InstanceDirName(), "kubeconfig")
```
Fwiw, I'd add a small helper such as:
```
func ResolveInstancePath(preset crcpreset.Preset, pathElems ...string) string {
	return filepath.Join(MachineInstancesDir, InstanceDirName(preset), filepath.Join(pathElems...))
}
```
to stop having  `filepath.Join(MachineInstanceDir, InstanceDirName(), someFile)` in many places.

cfergeau commented on pkg/crc/oc/oc.go:
```diff
@@ -81,7 +81,7 @@ func UseOCWithSSH(sshRunner *ssh.Runner) Config {
 		KubeconfigPath:   "/opt/kubeconfig",
 		Context:          constants.DefaultContext,
-		Cluster:          constants.DefaultName,
+		Cluster:          constants.DefaultClusterName,
```
seeing https://github.com/crc-org/crc/blob/233df0ee1c05f91d49dbe72dc58a2288ac09f30f/pkg/crc/machine/start.go#L556-L559, this could be further improved in follow-up commits.

cfergeau commented on pkg/crc/constants/constants.go:
```diff
@@ -174,24 +174,24 @@ func EnsureBaseDirectoriesExist() error {
 
 func GetPublicKeyPath() string {
-	return filepath.Join(MachineInstanceDir, InstanceDirName(), "id_ecdsa.pub")
+	return filepath.Join(MachineInstancesDir, InstanceDirName(), "id_ecdsa.pub")
```
I'd prefer to focus for now on having VM names, or machine dir name being `crc-$preset` instead of just `crc`, but keep for now "only one VM can exist at a given time".
This means most of this PR gets in, we can start making sure the change of names don't cause too many issues, ...
Then another PR can add support for multiple instance dir in the same subdir, multiple VMs, ...

cfergeau commented on pkg/crc/constants/constants.go:
```diff
@@ -113,12 +113,12 @@ var (
 
-func InstanceDirName() string {
-	return InstanceName()
+func InstanceDirName(preset crcpreset.Preset) string {
```
After this change, compilation will be broken for a few commits. Since it is a very mechanical change, doing everything in one go should be ok.

cfergeau commented on pkg/crc/preflight/preflight_checks_linux.go:
```diff
@@ -577,46 +577,50 @@ func removeLibvirtCrcNetwork() error {
-	}
-	if strings.TrimSpace(stdout) == "running" {
-		_, stderr, err := crcos.RunWithDefaultLocale("virsh", "--connect", "qemu:///system", "destroy", constants.InstanceName())
+	for _, instanceName := range getAllInstanceNamesToDelete() {
```
I'd prefer to see this commit split in 2 parts, one adjusting for the addition of the `preset` arg to the `constants.InstanceName()` function, and another one adding the iteration over multiple directories. 
Not sure we want the iteration over multiple dirs right now.

cfergeau replied:
This commit is also doing a lot more than what one might think by reading the commit log.

anjannath replied:
> I'd prefer to see this commit split in 2 parts, one adjusting for the addition of the `preset` arg to the `constants.InstanceName()` function, and another one adding the iteration over multiple directories. 

that'd be tricky, as preflight checks don't have a way to know the current preset, we could pass the `config.Storage` to each preflight check.. but i added the iteration over machine dirs part to avoid this

> Not sure we want the iteration over multiple dirs right now.

if we don't allow more then one preset to run then we don't need to delete all the machine dirs right now, but this was a workaround for the above problem as well, this way we don't have to change the signature of `checkFunc` and since we'll need it later when having multiple presets running, i'd prefer it to add this now itself

cfergeau replied:
> that'd be tricky, as preflight checks don't have a way to know the current preset, we could pass the config.Storage to each preflight check.. 

There are already a few places where this is done. I'm not a huge fan of how it's done, but this works, see https://github.com/cfergeau/crc/commit/aec4e14c43f0314a33218feb4f909a4ac188d838

cfergeau commented on cmd/crc/cmd/start.go:
```diff
@@ -346,7 +346,6 @@ func writeMicroShiftTemplatedMessage(writer io.Writer) error {
 	return parsed.Execute(writer, &templateVariables{
 		EvalCommandLine:   shell.GenerateUsageHint(userShell, "crc oc-env"),
 		CommandLinePrefix: commandLinePrefix(userShell),
-		KubeConfigPath:    constants.KubeconfigFilePath,
```
I guess this is unused? Might deserve its own commit/explanation.

cfergeau commented on pkg/crc/machine/virtualmachine.go:
```diff
@@ -101,10 +103,23 @@ func (vm *virtualMachine) SSHPort() int {
+		return preset.Preset(""), err
+	}
+	return parsedPreset, nil
+}
```
hardcoding a vm name -> preset mapping outside of constants.go (and even doing this) is something I'd prefer to avoid.

anjannath replied:
dropped, this was only needed in one place, and https://github.com/crc-org/crc/pull/3750/commits/3517405b7c03a55012adcadd06eb4a83b13ebaf1#r1317265025 solved that, thanks!

cfergeau commented on pkg/crc/machine/virtualmachine.go:
```diff
@@ -101,10 +103,23 @@ func (vm *virtualMachine) SSHPort() int {
+	if err != nil {
+		return nil, err
+	}
+	return ssh.CreateRunner(ip, vm.SSHPort(), constants.GetPrivateKeyPath(preset), constants.GetRsaPrivateKeyPath(), vm.bundle.GetSSHKeyPath())
```
`vm.bundle.GetBundleType()` will tell you which preset to use.

cfergeau commented on pkg/crc/constants/constants.go:
```diff
@@ -121,6 +120,13 @@ func InstanceName(preset crcpreset.Preset) string {
+	if preset != crcpreset.Podman {
+		return filepath.Join(MachineInstancesDir, InstanceDirName(preset), "kubeconfig")
+	}
+	return ""
```
podman was not special cased before, why the change? Should this go to a commit of its own maybe?

anjannath replied:
podman doesn't require this helper, i've removed the check and kept it like it was before

cfergeau commented on pkg/crc/api/api_http.go:
```diff
@@ -9,7 +9,7 @@ import (
 
 func NewMux(config *crcConfig.Config, machine machine.Client, logger Logger, telemetry Telemetry) http.Handler {
-	handler := NewHandler(config, machine, logger, telemetry)
+	handler := NewHandler(config, machine, logger, telemetry, MachineClientUpdater)
```
This commit seems to be about supporting multiple VMs running at the same time? As in the previous commits, I'd prefer to keep this kind of changes out of this PR, there's already a lot to get right without adding this kind of complication.

anjannath replied:
> This commit seems to be about supporting multiple VMs running at the same time?

this is needed just for having different VM names, not necessarily enabling them to run at the same time.

since we pass the machine.Client to this function, it is tied to one _VM name_, but when the preset is changed, we need a new machine.Client with the new preset's VM name, earlier when all the VMs were named `crc` this was not needed

cfergeau commented on test/integration/resize_test.go:
```diff
@@ -49,20 +50,20 @@ var _ = Describe("vary VM parameters: memory cpus, disk", Label("openshift-prese
 
 		It("check VM's memory size", func() {
-			out, err := SendCommandToVM("cat /proc/meminfo")
+			out, err := SendCommandToVM("cat /proc/meminfo", preset.OpenShift)
```
`SendCommandToVM` does not really need a preset arg as it's hardcoded to `OpenShift`. In my opinion, it would make more sense to add a constant equal to `preset.OpenShift` and use this instead. OpenShift will be hardcoded in a single place, and won't be hardcoded in SendCommandToVM, which I think is what you want to avoid.

anjannath replied:
`SendCommandToVM` is generic, it needs the preset to call `constants.InstanceName()` to get the VM name, and get the connection details for it, it can be used for the microshift preset with `SendCommandToVM("cat /proc/meminfo", preset.MicroShift)`.

i am not sure if it'd make more sense to use something like:
```
preset := preset.OpenShift
out, err := SendCommandToVM("cat /proc/meminfo", preset)
```
or did i completely misunderstood the comment?

anjannath replied:
i see since this preset is being used in all the invocation, it does make sense to use the above ^^

cfergeau reviewed (commented):
Finally a first round of comments! Overall looks good to me, though I'd prefer that for now we focus on the renaming, having VMs or instance dir being named `crc-$preset` instead of `crc`, and make sure this works fine. Having multiple instance dirs at once, or multiple VMs running at the same time should be done in other PRs, as this adds even more complication.

Regarding the current PR, does it work to create a microshift VM with a released version of crc, stop it, upgrade to a version of crc with the changes from this PR, and then start it again?

What about `crc start; crc config set preset microshift; crc start`? Does it still have the same behaviour as before the PR? I haven't tried to run your changes yet.

anjannath committed:
remove unused member KubeConfigPath from start msg template struct

anjannath committed:
oc: create oc config for ocp or microshift based on the preset

this modifies oc.UseOCWithSSH to also take the preset and return
oc config for either ocp or microshift based on the given preset

cfergeau commented:
I've rebased this.

anjannath commented:
> Regarding the current PR, does it work to create a microshift VM with a released version of crc, stop it, upgrade to a version of crc with the changes from this PR, and then start it again?
> 
no, since the older VM will be named as `crc` and the upgraded crc will look for a VM named `crc-microshift` it'll think machine does not exists

> What about `crc start; crc config set preset microshift; crc start`? Does it still have the same behaviour as before the PR? I haven't tried to run your changes yet.

No for this as well, it will  not warn the user that an existing VM with different preset already exists like it does now

anjannath committed:
start: prevent starting a preset when VM with another preset exists

this adds a check to prevent starting a new preset when another preset
already has a VM, this helps in keeping the start behavior same as  it
currently exists

anjannath commented:
closing as it is hard to keep the old behavior and still allow the VMs to be named after the presets

e.g
```
$ crc start (default preset)
$ crc config set preset microshift (this'll ask user to delete the old preset and start a new one)
$ crc delete (this will try to delete the non-existent microshift preset)
```

cfergeau commented:
For this specific usecase, we could block preset change if there's an existing VM, and improve teh behaviour later on
